safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,208 @@
+	BaseURI string
+	// ReportURI controls the report-uri directive. If ReportUri is empty, no report-uri
+	// directive will be set.
+	ReportURI string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		reportOnly: s.ReportOnly,
+		serialize: func(nonce string) string {
+			var b strings.Builder
+
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			b.WriteString(nonce)
+			b.WriteString("'")
+
+			if s.StrictDynamic {
+				b.WriteString(" 'strict-dynamic'")
+			}
+			if s.UnsafeEval {
+				b.WriteString(" 'unsafe-eval'")
+			}
+
+			b.WriteString("; base-uri ")
+			if s.BaseURI == "" {
+				b.WriteString("'none'")
+			} else {
+				b.WriteString(s.BaseURI)
+			}
+
+			if s.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(s.ReportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// FramingPolicyBuilder can be used to create a new CSP policy with frame-ancestors
+// set to 'self'.
+//
+// TODO: allow relaxation on specific endpoints according to #77.
+type FramingPolicyBuilder struct {
+	// ReportOnly controls whether this policy should be set as a Content-Security-Policy
+	// header or a Content-Security-Policy-Report-Only header.
+	ReportOnly bool
+	// ReportURI controls the report-uri directive. If ReportUri is empty, no report-uri
+	// directive will be set.
+	ReportURI string
+}
+
+// Build creates a Policy based on the specified options.
+func (f FramingPolicyBuilder) Build() Policy {
+	return Policy{
+		reportOnly: f.ReportOnly,
+		serialize: func(_ string) string {
+			var b strings.Builder
+			b.WriteString("frame-ancestors 'self'")
+
+			if f.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(f.ReportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	Policies []Policy
+}
+
+// NewInterceptor creates an interceptor from the provided policies.
+func NewInterceptor(p ...Policy) Interceptor {
+	return Interceptor{Policies: p}
+}
+
+// Default creates a new CSP interceptor with a strict nonce-based policy and a
+// framing policy, both in enforcement mode.
+func Default(reportURI string) Interceptor {
+	return Interceptor{
+		Policies: []Policy{
+			StrictCSPBuilder{ReportURI: reportURI}.Build(),
+			FramingPolicyBuilder{ReportURI: reportURI}.Build(),
+		},
+	}
+}
+
+// Before claims and sets the Content-Security-Policy header and the
+// Content-Security-Policy-Report-Only header.
+func (it Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	var csps []string
+	var reportCsps []string
+	nonce := generateNonce()
+	r.SetContext(context.WithValue(r.Context(), ctxKey{}, nonce))
+	for _, p := range it.Policies {
+		v := p.serialize(nonce)
+		if p.reportOnly {
+			reportCsps = append(reportCsps, v)
+		} else {
+			csps = append(csps, v)
+		}
+	}
```

Comment:
Can we somehow do this work only once?

Reply:
I changed it so that an `Interceptor` contains two lists of policies. One which is always enforced and one which is always report-only.